flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -51,7 +52,8 @@
  */
 @PublicEvolving
 public abstract class AsyncSinkWriter<InputT, RequestEntryT extends Serializable>
-        implements StatefulSink.StatefulSinkWriter<InputT, BufferedRequestState<RequestEntryT>> {
+        implements StatefulSink.StatefulSinkWriter<InputT, BufferedRequestState<RequestEntryT>>,
+                TwoPhaseCommittingSink.PrecommittingSinkWriter<InputT, Void> {
```

Comment:
yes, for the current state `SinkWriterOperator` calls `prepareCommit` for all `PrecommittingSinkWriter`s even the ones via a non `TwoPhaseCommittingSink` Sinks. It is stated [here](https://github.com/apache/flink/blob/dcd215620d96c5e648cea777f0b7594f7b5a1e26/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java#L178) here that it is to keep a sink v1 topology behaviour.
It is called upon calling [prepareSnapshotPreBarrier](https://github.com/apache/flink/blob/dcd215620d96c5e648cea777f0b7594f7b5a1e26/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java#L137) which I understand is called directly before checkpointing.

Reply:
After a second look.
I believe this issue should be resolved, as per this [line](https://github.com/apache/flink/blob/dcd215620d96c5e648cea777f0b7594f7b5a1e26/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java#L147)
the `flush` method is used to notify the writer with the upcoming snapshot.

It appears that the `prepareCommit` interface is only used in 2PC cases and for operators that would separate building the immutable state object and returning it in `snapshotState`. Since `AsyncSinkWriter` creates the state object on returning it, I think this PR is not needed.

Reply:
But `flush` is not blocking so after calling `flush` there will still be requests in flight, right? Am I missing something?

Reply:
I thought it is blocking now after merging #18651.

Reply:
Ah yes I see, sorry I was looking at the wrong flush, it is overloaded

Reply:
https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java#L494